push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,122 @@
+
+## Problem Description
+
+Rather than highlighting problems, User Stories will be used to capture improvements and value that can be added that makes for a better overall experience when using UPS on OpenShift
+
+*As a developer, I want UPS running in my OpenShift namespace, ready to go without any additional setup so I can start to use it immediately*
+
+The Unified Push Server is made up of a Java application & a database (MySQL).
+It also has a dependecy on an auth provider (currently Keycloak) for users of the UPS Administration UI or REST API.
+There is a proof of concept UPS Ansible Playbook Bundle (APB) available.
+However, the APB just provisions UPS and MySQL currently.
+It expects an existing Keycloak instance to be running somewhere with the appropriate Realm, Client & Roles already setup.
+This can be improved to eliminate any additional setup or prerequisites required from the developer.
+
+*As a developer, I want a consistent login experience across all Mobile Services so that I can easily login with the same credentials to administer my services*
+
+As detailed in [Sign On Proposal](../auth/developer-single-sign-on-across-mobile-services.md), a goal with all Mobile Services is to have a consistent login experience by logging in using OpenShift credentials.
+This currently isn't the case as a separate User in Keycloak is used.
+
+*As a developer, I want a simplified setup of Push variants in UPS so that I don't have to manually define them and link them back to Mobile Clients in OpenShift*
+
+Currently there is a switch in context and terms between creating Mobile Clients in OpenShift, and setting up Push Applications & Variants in UPS.
+The developer creates a 'Mobile Client' in OpenShift as a logical construct to represent the App they are developing outside of OpenShift.
+In UPS, the developer creates a 'Push Application' and 1 or more 'Variants' to represent the App they are developing outside of OpenShift.
+There is duplication of logical constructs and potential confusion between what Mobile Client is and what a Push Application is.
+
+The developer also has to somehow get the Push configuration from UPS and place it into their App for consumption by the SDK. With the mobile-services.json approach being used by the Mobile CLI/UI and the SDKs for configuration, there's some things that can be done to simplify what the developer has to do here.
+
+*As a developer, I want to write custom logic for sending Push messages so that I can dynamically send messages based on varying factors and inputs*
+
+Push messages can be sent by calling the [sender API](https://aerogear.org/docs/specs/aerogear-unifiedpush-rest/index.html#397083935) with the correct key. There are features available in the OpenShift and the Catalog that can simplify the provisoining of a custom service, and obtaining secret credentials (like a key) from a provisioned service.
```

Comment:
How will the developer get the `PushApplicationID:MasterSecret` to send the messages?

Reply:
When the developer 'binds' to UPS, a secret will be created that has the master UPS secret value.
This secret can be mounted into a custom service or set as an env var for a custom service.